pkg/socketservice/socketservice.go

```diff
@@ -179,6 +183,128 @@ func processDeploySocketForApp(clusterSocket *ClusterSocket, a *apptypes.App) (b
 	return true, nil
 }
 
+func kustomizeCharts(deployedVersionArchive string, name string, version string) ([]byte, error) {
+	archive := []byte{}
+	archiveChartDir := filepath.Join(deployedVersionArchive, "overlays", "downstreams", name, "charts")
+	_, err := os.Stat(archiveChartDir)
+	if err != nil && os.IsNotExist(err) {
+		return archive, nil
+	}
+	if err != nil && !os.IsNotExist(err) {
+		return archive, nil
+	}
+
+	exportChartPath, err := ioutil.TempDir("", "kotsadm")
+	if err != nil {
+		return archive, errors.Wrap(err, "failed to create temp dir")
+	}
+	defer os.RemoveAll(exportChartPath)
+	exportChartsDir := filepath.Join(exportChartPath, "charts")
+	if _, err := os.Stat(exportChartsDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(exportChartsDir, 0744); err != nil {
+			return archive, errors.Wrap(err, "failed to mkdir for archive chart")
+		}
+	}
+
+	baseDir := filepath.Join(deployedVersionArchive, "base", "charts")
+	baseFiles := []string{"Chart.yaml", "Chart.lock"}
+
+	err = filepath.Walk(archiveChartDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			rel, err := filepath.Rel(archiveChartDir, filepath.Dir(path))
+			if err != nil {
+				return err
+			}
+			for _, filename := range baseFiles {
+				err = exportFile(baseDir, rel, exportChartsDir, filename)
+				if err != nil {
+					return errors.Wrap(err, "failed to export file")
+				}
+			}
+			if info.Name() == "kustomization.yaml" {
+				archiveChartOutput, err := exec.Command(fmt.Sprintf("kustomize%s", version), "build", filepath.Dir(path)).Output()
+				if err != nil {
+					if ee, ok := err.(*exec.ExitError); ok {
+						err = fmt.Errorf("kustomize stderr: %q", string(ee.Stderr))
+					}
+					return errors.Wrap(err, "failed to kustomize")
+				}
+				err = exportContent(archiveChartOutput, rel, exportChartsDir, "all.yaml")
+				if err != nil {
+					return errors.Wrap(err, "failed to export content")
+				}
+			}
+			return nil
+		})
+	if err != nil {
+		return archive, err
+	}
+
+	tempDir, err := ioutil.TempDir("", "helmkots")
+	if err != nil {
+		return archive, errors.Wrap(err, "failed to create temp dir")
+	}
+	defer os.RemoveAll(tempDir)
+	tarGz := archiver.TarGz{
+		Tar: &archiver.Tar{
+			ImplicitTopLevelFolder: true,
+		},
+	}
+	if err := tarGz.Archive([]string{exportChartsDir}, path.Join(tempDir, "helmcharts.tar.gz")); err != nil {
+		return archive, errors.Wrap(err, "failed to create tar gz")
+	}
+	archive, err = ioutil.ReadFile(path.Join(tempDir, "helmcharts.tar.gz"))
+	if err != nil {
+		return archive, errors.Wrap(err, "failed to read helm tar.gz file")
+	}
+	return archive, nil
+}
+
+func exportContent(allContent []byte, rel string, exportChartsDir string, filename string) error {
+	relDir := ""
+	if filepath.Base(rel) != "crds" {
+		relDir = filepath.Join(exportChartsDir, rel, "templates")
```

Comment:
if its the case that this is required, we should maintain this in the work i did with the file tree

Reply:
this is to reconstruct the chart. i dont think we need to maintain the work in the file tree. we can in future to optimize may be.